jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceCollection.java

```diff
@@ -245,59 +233,54 @@ public Resource addPath(String path) throws IOException
 
         if (path == null)
         {
-            throw new MalformedURLException();
+            throw new MalformedURLException("null path");
         }
 
         if (path.length() == 0 || URIUtil.SLASH.equals(path))
         {
             return this;
         }
 
-        Resource resource = null;
-        ArrayList<Resource> resources = null;
-        int i = 0;
-        for (; i < _resources.length; i++)
+        // Attempt a simple (single) Resource lookup that exists
+        for (Resource res : _resources)
         {
-            resource = _resources[i].addPath(path);
-            if (resource.exists())
+            Resource fileResource = res.addPath(path);
+            if (fileResource.exists())
             {
-                if (resource.isDirectory())
+                if (!fileResource.isDirectory())
                 {
-                    break;
+                    return fileResource;
                 }
-                return resource;
             }
         }
 
-        for (i++; i < _resources.length; i++)
+        // Create a list of potential resource for directories of this collection
+        ArrayList<Resource> potentialResources = null;
+        for (Resource res : _resources)
         {
-            Resource r = _resources[i].addPath(path);
+            Resource r = res.addPath(path);
             if (r.exists() && r.isDirectory())
             {
-                if (resources == null)
+                if (potentialResources == null)
                 {
-                    resources = new ArrayList<>();
+                    potentialResources = new ArrayList<>();
                 }
 
-                if (resource != null)
-                {
-                    resources.add(resource);
-                    resource = null;
-                }
-
-                resources.add(r);
+                potentialResources.add(r);
             }
         }
 
-        if (resource != null)
+        if (potentialResources == null || potentialResources.isEmpty())
         {
-            return resource;
+            throw new MalformedURLException("path does not result in Resource: " + path);
         }
-        if (resources != null)
+
+        if (potentialResources.size() == 1)
```

Comment:
This substantially changes the semantic:  before we only ever returned null, or  the first matching resource; now, you throw instead of null (fine) but only return the first resource if there is only one, otherwise you return the whole lot. Pretty sure there's code that's not going to deal with that. Can't we leave the semantic as it was before, ie only return the first?

Also, I don't love the var name "potentialResources", what was wrong with "resources"?